### DISCUSSION TITLE
[archived] "Workflow Pod Security Context" docs discussion

### DISCUSSION THREAD

**Discussion by giscus** [argoproj/argo-workflows#12983] [General]
# en/latest/workflow-pod-security-context/

https://argo-workflows.readthedocs.io/en/latest/workflow-pod-security-context/

**Comment by joebowbeer** [comment]
pod security policy was removed in k8s 1.25 and replaced with pod and container security context.

**Reply by agilgur5** [reply]
> in k8s 1.25 and replaced with pod and container security context.

Pod and container `securityContext` have existed much, much longer than that. That is what this page is about as well, which also predates 1.25 by years.

[PSPs](https://kubernetes.io/docs/concepts/security/pod-security-policy/) were an _enforcement mechanism_ that were not very intuitive and so were replaced with [PSA](https://kubernetes.io/docs/concepts/security/pod-security-admission/) / [PSS](https://kubernetes.io/docs/concepts/security/pod-security-standards) / policy controllers.

None of those change usage of `securityContext`, but they may change _enforcement_ of `securityContext`

**Comment by joebowbeer** [comment]
@agilgur5 My comment refers to the following, which even links to the PSP docs:

> This is likely to be necessary if you have a [pod security policy](https://kubernetes.io/docs/concepts/policy/pod-security-policy/). You probably can't use the Docker executor if you have a pod security policy.

**Reply by agilgur5** [reply]
Yes I am aware. The PSP docs also themselves say that.

I was specifically [responding](https://github.com/argoproj/argo-workflows/discussions/12983#discussioncomment-9238932) to the part of your comment saying:
> replaced with pod and container security context.

which is not accurate (nor what the PSP docs say).

<hr />

The docs line is still accurate; if you're using PSPs, PSA, or similar, it is likely still necessary to use `securityContext`.
Argo also supports versions of k8s prior to 1.25, which was only EoL 6 months ago and is still supported by certain vendors, e.g. EKS, so it is still relevant. If you'd like to update the line to include PSA and similar, PR welcome.

The Docker executor part is outdated, see #12975

**Reply by joebowbeer** [reply]
Restating my comment as a suggestion:

_Mention of PSP, which was removed in k8s 1.25, should be removed from this page._

PSP has been deprecated since k8s 1.21 and was last seen in k8s 1.24, which reached EOL 9 months ago (28 Jul 2023).

Though it [looks like](https://argo-workflows.readthedocs.io/en/stable/releases/#kubernetes-compatibility-matrix) (?) Argo Workflows 3.4 might still support k8s 1.24 (really?)

In any event, I think this unqualified mention of something that no longer exists is confusing, which is why I added this comment.

I suggest either removing the mention or adding a qualifier such as "which has been deprecated since Kubernetes 1.21".

As for the statement that PSP has been replaced by pod and container security contexts, this was made from the perspective of the security benchmarks and best practices. Shortly after PSP was deprecated, the existing references to PSP were replaced with guidelines for pod and container security context.

**Reply by agilgur5** [reply]
> Though it [looks like](https://argo-workflows.readthedocs.io/en/stable/releases/#kubernetes-compatibility-matrix) (?) Argo Workflows 3.4 might still support k8s 1.24 (really?)

It does. In fact, 3.4 might even support lower than that -- part of the test matrix [tests k8s 1.21](https://github.com/argoproj/argo-workflows/blob/3.4.16/.github/workflows/ci-build.yaml#L90).

> something that no longer exists

They do still exist for many users. [EKS still supports](https://docs.aws.amazon.com/eks/latest/userguide/kubernetes-versions.html#available-versions-extended) 1.23 and even 1.21. I literally spoke to a user on Slack with a PSP error [yesterday](https://cloud-native.slack.com/archives/C01QW9QSSSK/p1714008213496449?thread_ts=1713875795.672269&cid=C01QW9QSSSK). I was explaining PSPs to a new k8s SRE team less than a year ago (kOps lags behind k8s, and the enterprise internal version we had at that job lagged even further behind. That's pretty common though unfortunately, there's much older software being supported all over, enough so to be a sizeable chunk of sales strategy of commercial extended support).

In general, we should keep around docs for longer than is strictly necessary as there will always be users on deprecated versions. There are still plenty of users on Argo 3.3 unfortunately too, which isn't supported either.
We also only landed versioned docs recently (#11390), so older things (including older Argo things) were also left in intentionally prior to that. I am removing some things per #12975, but will still leave them in older versions of the docs, even ones that no longer support the feature.
Broadly speaking, as a general rule of thumb, leave docs for something that is 1-2 versions behind support.

> or adding a qualifier

Yes I think this would be more of a middle ground between different demographics of users, each who may be more or less confused by a wholesale removal.
We can use the word "legacy" or "deprecated". It would also make sense to include PSA, PSS, et al, as I wrote above.

> Shortly after PSP was deprecated, the existing references to PSP were replaced with guidelines for pod and container security context.

I'm pretty sure you're referring to [PSS](https://kubernetes.io/docs/concepts/security/pod-security-standards/). `securityContext`s were still best practice before them, PSS just introduced standards and terminology that could be _enforced_ on them by k8s itself (via PSA) and other tools. PSS basically formalized what the default PSPs already enforced.

**Reply by joebowbeer** [reply]
@agilgur5 #12990

**Reply by agilgur5** [reply]
Thanks @joebowbeer!
